### PR TITLE
Ajax: Don't mangle the URL when removing the anti-cache param

### PR DESCRIPTION
**Pull request by dmethvin**
### Summary ###

Fixes gh-3229.

There was an existing unit test that actually showed the URL being mangled but there were not any checks for the problem.

### Checklist ###
Mark an `[x]` for completed items, if you're not sure leave them unchecked and we can assist.

* [x] All authors have signed the CLA at https://contribute.jquery.com/CLA/
* [x] New tests have been added to show the fix or feature works
* [x] Grunt build and unit tests pass locally with these changes
* [x] ~~If needed, a docs issue/PR was created at https://github.com/jquery/api.jquery.com~~

Thanks! Bots and humans will be around shortly to check it out.

### CODE REVIEW

File: src/ajax.js [hunk 1]
```diff
@@ -18,7 +18,8 @@ define( [
 var
 	r20 = /%20/g,
 	rhash = /#.*$/,
-	rts = /([?&])_=[^&]*/,
+	rantiCache = /(?:[?&])_=[^&]*&?/,
```
Comment (mgol) [inline src/ajax.js]: Sounds almost as "ranty cache". ;-)

Comment (mgol) [inline src/ajax.js]: What is the first group for?

Comment (dmethvin) [inline src/ajax.js]: Oh good catch, that was from an earlier failed attempt to solve the problem without having to make a second pass to remove the empty `?` querystring. I don't think its needed now.

File: src/ajax.js [hunk 2]
```diff
@@ -597,15 +598,16 @@ jQuery.extend( {
...
 
 				// #9682: remove data so that it's not used in an eventual retry
 				delete s.data;
 			}
 
-			// Add anti-cache in uncached url if needed
+			// Add or update anti-cache param if needed
 			if ( s.cache === false ) {
-				cacheURL = cacheURL.replace( rts, "" );
+				cacheURL = cacheURL.replace( rantiCache, "" ).replace( remptyQuery, "" );
```
Comment (gibson042) [inline src/ajax.js]: This code still exhibits the issue (e.g., it converts `/path?_=1&foo=bar` into `/pathfoo=bar`). The root cause of gh-3229 isn't a bad regex, it's a missing "$1" replacement string dropped by https://github.com/jquery/jquery/commit/e077ffb083743f4a4b990f586c9d25d787e7b417#diff-c3749d3acba09ca9ec16bb56e496408bL599 that we just never noticed.

File: src/ajax.js [hunk 3]
```diff
@@ -18,7 +18,8 @@ define( [
 var
 	r20 = /%20/g,
 	rhash = /#.*$/,
-	rts = /([?&])_=[^&]*/,
+	rantiCache = /(?:[?&])_=[^&]*&?/,
+	remptyQuery = /\?$/,
```
Comment (gibson042) [inline src/ajax.js]: I wouldn't worry about this at all, since `/path?&_=1` is valid. But if we do care, then it can be solved by restricting the existing regex capture group rather than introducing a new regex: `/(?:(\?)|&)_=[^&]*/` or (probably smaller) `/(\?)_=[^&]*|&_=[^&]*/`. It should be a separate PR, though.

File: src/ajax.js [hunk 4]
```diff
@@ -597,15 +598,16 @@ jQuery.extend( {
 
 			// If data is available, append data to url
 			if ( s.data ) {
-				cacheURL += ( rquery.test( cacheURL ) ? "&" : "?" ) + s.data;
+				cacheURL = cacheURL.replace( remptyQuery, "" ) +
+					( rquery.test( cacheURL ) ? "&" : "?" ) + s.data;
```
Comment (gibson042) [inline src/ajax.js]: If we really want to update the "append parameter" logic to handle terminal `?`s, I'd rather see something like `(cacheURL + "&").replace( /\?(.+&$)|\??&$/, "?$1" )` (dropping `rquery` in favor of the new regex) in a separate PR also including [src/ajax/jsonp.js#L47](https://github.com/jquery/jquery/blob/305f193aa57014dc7d8fa0739a3fefd47166cd44/src/ajax/jsonp.js#L47).

**Comment by dmethvin** [comment]
Thanks to @gibson042 I think I've fixed the real problem here. The unit test improvements that verify the URL will hopefully prevent a regression.

**Comment by gibson042** [comment]
Works for me.
